File: dsa-problems/StringsClassic/solutions/javascript/STC-001-kmp-prefix-function.js
const readline = require("readline");

class Solution {
  prefixFunction(s) {
    const n = s.length;
    const pi = new Array(n).fill(0);
    let j = 0; // length of the previous longest prefix
    
    for (let i = 1; i < n; i++) {
      // Backtrack
      while (j > 0 && s[i] !== s[j]) {
        j = pi[j - 1];
      }
      // Extend
      if (s[i] === s[j]) {
        j++;
      }
      pi[i] = j;
    }
    return pi;
  }
}

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

let data = [];
rl.on("line", (line) => {
  const parts = line.trim().split(/\s+/);
  for (const part of parts) {
    if (part) data.push(part);
  }
});

rl.on("close", () => {
  if (data.length === 0) return;
  const s = data[0];
  const solution = new Solution();
  const pi = solution.prefixFunction(s);
  console.log(pi.join(" "));
});

File: dsa-problems/StringsClassic/solutions/cpp/STC-001-kmp-prefix-function.cpp
#include <iostream>
#include <vector>
#include <string>

using namespace std;

class Solution {
public:
    vector<int> prefixFunction(const string& s) {
        int n = s.length();
        vector<int> pi(n, 0);
        int j = 0; // length of the previous longest prefix
        
        for (int i = 1; i < n; i++) {
            // Backtrack
            while (j > 0 && s[i] != s[j]) {
                j = pi[j - 1];
            }
            // Extend
            if (s[i] == s[j]) {
                j++;
            }
            pi[i] = j;
        }
        return pi;
    }
};

int main() {
    // Fast I/O
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    string s;
    if (cin >> s) {
        Solution solution;
        vector<int> pi = solution.prefixFunction(s);
        
        for (int i = 0; i < (int)pi.size(); i++) {
            if (i > 0) cout << " ";
            cout << pi[i];
        }
        cout << "\n";
    }
    return 0;
}
